system-addon/content-src/components/TopSites/TopSiteForm.jsx

```diff
@@ -84,16 +152,61 @@ export class TopSiteForm extends React.PureComponent {
     }
   }
 
+  validateCustomScreenshotUrl() {
+    const {customScreenshotUrl} = this.state;
+    return customScreenshotUrl ? this.validateUrl(customScreenshotUrl) : true;
+  }
+
   validateForm() {
-    const validate = this.validateUrl(this.state.url);
-    this.setState({validationError: !validate});
+    const validate = this.validateUrl(this.state.url) && this.validateCustomScreenshotUrl();
+
+    if (!validate) {
+      this.setState({validationError: true});
+    }
+
     return validate;
   }
 
+  _renderCustomScreenshotInput() {
+    const validationError = this.state.validationError &&
+      (this.state.screenshotRequestFailed || !this.validateCustomScreenshotUrl());
+    // Set focus on error if the url field is valid or when the input is first rendered and is empty
+    const shouldFocus = (validationError && this.validateUrl(this.state.url)) || !this.state.customScreenshotUrl;
+
+    if (!this.state.showCustomScreenshotForm) {
+      return (<a href="" className="enable-custom-image-input" onClick={this.onEnableScreenshotUrlForm}>
+        <FormattedMessage id="topsites_form_use_image_link" />
+      </a>);
+    }
+    return (<div className="custom-image-input-container">
+      <TopSiteFormInput
+        errorMessageId={this.props.screenshotRequestFailed ? "topsites_form_image_validation" : "topsites_form_url_validation"}
+        loading={this.state.pendingScreenshotUpdate}
+        onChange={this.onCustomScreenshotUrlChange}
+        onClear={this.onClearScreenshotInput}
+        shouldFocus={shouldFocus}
+        typeUrl={true}
+        value={this.state.customScreenshotUrl}
+        validationError={validationError}
+        titleId="topsites_form_image_url_label"
+        placeholderId="topsites_form_url_placeholder"
+        intl={this.props.intl} />
+    </div>);
+  }
+
+  _getPreviewScreenshot() {
+    return Object.assign({}, this.props.site, {screenshotPreview: this.state.screenshotPreview});
+  }
+
   render() {
+    const {customScreenshotUrl} = this.state;
     // For UI purposes, editing without an existing link is "add"
     const showAsAdd = !this.props.site;
-
+    const changedCustomScreenshotUrl = customScreenshotUrl &&
+      (showAsAdd || (this.props.site.customScreenshotURL !== this.cleanUrl(customScreenshotUrl)));
+    // Preview mode enables the preview button and prevents saving or adding a topsite.
+    const previewMode = (changedCustomScreenshotUrl && !this.state.screenshotPreview) ||
+      this.state.screenshotRequestFailed;
```

Comment:
Probably just have the comment describe the condition we're trying to match especially because it seems like the logic had to be fixed up here a few times to get right, but I'm not even sure if this is right?

Reply:
In particular, I would have expected some check for pending? And pending should be set only if we haven't received a preview/failure for the most recently requested url?

Reply:
>  I would have expected some check for pending

You have to be in `PreviewMode` in order to trigger a request and get to pending. Unless you meant something else.